### ISSUE TITLE
Diff chunks are not minimal

### ISSUE THREAD

**Issue by bersbersbers** [microsoft/vscode#131091]
[Monace Playground Repro](https://microsoft.github.io/monaco-editor/playground.html?source=v0.40.0-dev.20230704#XQAAAALKAgAAAAAAAABBqQkHQ5NjdMjwa-jY7SIQ9S7DNlzs5W-mwj0fe1ZCDRFc9ws9XQE0SJE1jc2VKxhaLFIw9vEWSxW3yscw2YmLaBJ5fGqgZNtIIGM6BnaWNlFk-HzucdJmwgffCJwWZ7P2mi3pjMKBhAZkCYzCwNyRuscN-jbeqHOmuKRscuPg32_4dMQJn6kkt0zCnND6-d5SdT_Dn5oV43PkAjaKYgK6QDgLw1eFA9w_Cc5pzPwKBt6KmxqhfUG3XcB4Wi4iZK7V0AGV0gxSk92451f2aneN1zh4G17jZOC1hNZnqtd1Wfavdf1c9fhWwH1RtTqTfDIoVFYOHmSCJvz8Io8H1OJUyoJkhPfRUvw5S3v-s4U5iVnErShK7kj9rtxt9qGR9Pty0fFB_iz3MbOAVprO5ryK5Wz9sPldv8ZIlitiul4WAety8AaKBU_Fp8gGio4tXLJH2_2gYHet-pgqhA8JXUTuo7M1wzgHcGRYqKuMn1OTk_LIfeD_p-ae6B-0_9TbUV4)

Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.59.0
- OS Version: Win 10

Steps to Reproduce:

`bug1.r`:
```r
{
    if (A) {
        if (B) {
            doit
        }
    }
}
C
D
```

`bug2.r`:
```r
{
    if (A && B) {
        doit
    }
}
C
E
```

Compare the two files:

![image](https://user-images.githubusercontent.com/12128514/129911893-9260238d-5fd2-4609-b79f-b5d714c24032.png)
![image](https://user-images.githubusercontent.com/12128514/129907032-f988ae87-aeb1-49f3-8912-b7fca6eb8d97.png)

Note how lines 6+7 are identical to lines 4+5 - there is no need for them to appear in the diff.

Change "E" to "D" in `bug2.r` and see the expected result:

![image](https://user-images.githubusercontent.com/12128514/129911950-0f165cc2-5c00-4074-b3de-8e6af16dcac7.png)
![image](https://user-images.githubusercontent.com/12128514/129907318-258a60c9-79d7-4c64-bcb1-0daae60a8b97.png)

Somehow, the second (D/E) diff chunk has an impact on the first diff chunk, which is not necessary at all.

**Comment by IllusionMH** [comment]
Looks like duplicate of #84981

**Comment by bersbersbers** [comment]
@IllusionMH well, other diff chunk issues have been solved quite quickly in the past, see https://github.com/microsoft/vscode/issues/119051 or https://github.com/microsoft/vscode/issues/121436.

**Comment by IllusionMH** [comment]
/assign @alexdima (a person who resolved 2 previous)

OK, wasn't aware of that changes.

**Comment by alexdima** [comment]
This is unfortunate, it has to do with how we compute diffs when trimming whitespace is not ignored: we actually compute the diff with trimming whitespace ignored and then introduce the char-level diffs for trimming whitespace. In this case, this backfires as there was a better `}` that could have been matched.

**Comment by hediet** [comment]
![chrome_Cf4dfZg7PJ](https://github.com/microsoft/vscode/assets/2931520/0ffd8b06-7a95-475f-b6d4-600eb27a473e)

Fixed. Try the monaco playground link in the first issue with the latest dev!

**Comment by bersbersbers** [comment]
@hediet I have two comments:

1. 

> ![chrome_Cf4dfZg7PJ](https://user-images.githubusercontent.com/2931520/258082489-0ffd8b06-7a95-475f-b6d4-600eb27a473e.png)

This screenshot seems to use `ignoreTrimWhitespace: true`. That is not what my issue was about, but I wonder if it was better if the original line 6 was deleted rather than original line 7, because the original line 7 is an exact match with the new line 5.

2. My original issue was about `ignoreTrimWhitespace: false, in which case I get this:

![image](https://github.com/microsoft/vscode/assets/12128514/36f4d46b-f732-4871-95aa-f1f67ec63a5b)

This looks pretty much like the snapshot I posted in the original comment. What was changed is that the CD/CE stuff does not impact the first chunk any longer - however, both cases (CD/CE and CD/CD) yield the non-minimal diff for the first chunk. So while the diff algorithm is more consistent now, its diffs tend to be longer than before. That is contrary of making them "minimal" ;)

**Comment by bersbersbers** [comment]
I am checking out 1.82.0, and I would expect this fix to be in there. But it seems it is not:

![image](https://github.com/microsoft/vscode/assets/12128514/9ecca964-8719-44c7-8f12-9d2867ca0926)

Line 5 should not be part of the diff IMHO.

**Comment by hediet** [comment]
Can you check if you use the legacy diff algorithm?

**Comment by bersbersbers** [comment]
> Can you check in your settings if you use the legacy diff algorithm?

Do you mean `diffEditor.diffAlgorithm`? I have tried both options (legacy and advanced), and I don't see a difference between the two. Sometimes, `diffEditor.ignoreTrimWhitespace` does not seem to have an effect, either:

![image](https://github.com/microsoft/vscode/assets/12128514/3d4400f7-1582-401e-b8d7-67bb364f9f3c)

![image](https://github.com/microsoft/vscode/assets/12128514/41775668-d132-4b7a-aa32-c7bb15c570c2)

I will investigate this separately - maybe my profile is broken somehow (or maybe not, see below).


In a new Profile (using the advanced algorithm), I *do* see a difference depending on `diffEditor.ignoreTrimWhitespace`:

![image](https://github.com/microsoft/vscode/assets/12128514/500c93f5-22c1-46b7-9692-aeb415f20b65)

![image](https://github.com/microsoft/vscode/assets/12128514/2b8a4c7b-00ea-464a-a05e-42c9c788514b)

After settings the diff algorithm to legacy, the result is similar, although now the highlights are gone and `diffEditor.ignoreTrimWhitespace` has stopped working again:

![image](https://github.com/microsoft/vscode/assets/12128514/362c5d21-4a05-4875-a618-36125b1ca46a)

This is the error I see in the console:
![image](https://github.com/microsoft/vscode/assets/12128514/40844315-3e3f-4ecc-abda-a406dec62e13)

After setting the algorithm back to advanced and reloading, the error messages don't appear any more, but `diffEditor.ignoreTrimWhitespace` is still/again without effect.

**Comment by bersbersbers** [comment]
> ![image](https://user-images.githubusercontent.com/12128514/266832124-2b8a4c7b-00ea-464a-a05e-42c9c788514b.png)

TLDR: advanced algorithm, with whitespaces: new line 5 should not be part of the diff chunk because it is identical to old line 7.

**Comment by bersbersbers** [comment]
Based on the [Playground](https://microsoft.github.io/monaco-editor/playground.html?source=v0.42.0-dev-20230805#XQAAAALKAgAAAAAAAABBqQkHQ5NjdMjwa-jY7SIQ9S7DNlzs5W-mwj0fe1ZCDRFc9ws9XQE0SJE1jc2VKxhaLFIw9vEWSxW3yscw2YmLaBJ5fGqgZNtIIGM6BnaWNlFk-HzucdJmwgffCJwWZ7P2mi3pjMKBhAZkCYzCwNyRuscN-jbeqHOmuKRscuPg32_4dMQJn6kkt0zCnND6-d5SdT_Dn5oV43PkAjaKYgK6QDgLw1eFA9w_Cc5pzPwKBt6KmxqhfUG3XcB4Wi4iZK7V0AGV0gxSk92451f2aneN1zh4G17jZOC1hNZnqtd1Wfavdf1c9fhWwH1RtTqTfDIoVFYOHmSCJvz8Io8H1OJUyoJkhPfRUvw5S3v-s4U5iVnErShK7kj9rtxt9qGR9Pty0fFB_iz3MbOAVprO5ryK5Wz9sPldv8ZIlitiul4WAety8AaKBU_Fp8gGio4tXLJH2_2gYHet-pgqhA8JXUTuo7M1wzgHcGRYqKuMn1OTk_LIfeD_p-ae6B-0_9TbUV4), I think this has been fixed in 0.42.0-dev-20230805 and is broken in 0.42.0-dev-20230804. Could it be that VS Code 1.82.0 ships with an older Monaco editor version?

**Comment by hediet** [comment]
> 1.82.0 ships with an older Monaco editor version?

That cannot be.

**Comment by bersbersbers** [comment]
Thanks! [This looks perfect!](https://microsoft.github.io/monaco-editor/playground.html?source=v0.44.0-dev-20230928#XQAAAALSAgAAAAAAAABBqQkHQ5NjdMjwa-jY7SIQ9S7DNlzs5W-mwj0fe1ZCDRFc9ws9XQE0SJE1jc2VKxhaLFIw9vEWSxW3yscw2cCzbyQm7h8f990-DBi5-c9Zhpkza7atELEWqIQxl445uEWLlBQR7lggfJc5jlyMnYRpg8foVlBH3g0V-NQV6Sgs4GzjQsAL0CshPG-Yph5QaYrKDn_uyS8e2pDk7GaxJf6prwPwOK7JFyZVTYBUKK4RhR0qD48nxvKBW-7vsiTEORc6J9IDBVXV9vu2Myt1XXDh0LgOIHJeLFOt3jr-Sdz_sVVed9sbmz4yduITIKzX82tinSmJEvFfluSLljkHrrJiXvKe2WT0we-1_pz9mgcuKBv-Tf8ybI8oCOrrXRdqMKkPo4ae3tYkTmi9HqCj0w5U_NZXmfhL4UgI0i6y-PLlZ4b4fEwn4iakCa2V-PYznJ4ze6nmhsH8ojj0voxsTgNYR5IyEMdg2BMC5eU80resv6-s2xkgAciQPhZaE3Eil6GU_9fP0S8)
